File: SumRootToLeafNumbers.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Tree
{
    class SumToLeafLC
    {
        //Time Complexity: O(n)
        //Space Complexity: O(h)
        //Definition of a binary tree node
        public class TreeNode
        {
            internal int val;
            internal TreeNode left;
            internal TreeNode right;
            TreeNode() { }
            TreeNode(int val) { this.val = val; }
            TreeNode(int val, TreeNode left, TreeNode right)
            {
                this.val = val;
                this.left = left;
                this.right = right;
            }
        }

        int sum;
        public int SumNumbers(TreeNode root)
        {
            if (root == null)
            {
                return 0;
            }
            helper(root, 0);
            return sum;
            //return helperWithLocalVariable(root, 0, 0); 
        }

        private void helper(TreeNode root, int currSum)
        {
            //base
            if (root == null)
            {
                return;
            }
            //logic
            currSum = currSum * 10 + root.val;
            helper(root.left, currSum);
            if (root.left == null && root.right == null)
            {
                sum += currSum;
            }
            helper(root.right, currSum);      
        }
        private int helperWithLocalVariable(TreeNode root, int currSum, int sum)
        {
            //base
            if (root == null)
            {
                return 0;
            }
            //logic
            currSum = currSum * 10 + root.val;
            if (root.left == null && root.right == null)
            {
                sum += currSum;
                return sum;
            }
            return helperWithLocalVariable(root.left, currSum, sum) + helperWithLocalVariable(root.right, currSum, sum);
        }

        class Pair
        {
            TreeNode node;
            int currSum;


            public Pair(TreeNode node, int currSum)
            {
                this.node = node;
                this.currSum = currSum;
            }

            public TreeNode getKey()
            {
                return this.node;
            }

            public int getValue()
            {
                return this.currSum;
            }
        }

        public int SumNumbersIteration(TreeNode root)
        {
            if (root == null)
            {
                return 0;
            }
            int sum = 0, currSum = 0;
            Stack<Pair> s = new Stack<Pair>();
            while (root != null || s.Count != 0)
            {
                while (root != null)
                {
                    currSum = currSum * 10 + root.val;
                    s.Push(new Pair(root, currSum));
                    root = root.left;
                }
                Pair p = s.Pop();
                root = p.getKey();
                currSum = p.getValue();
                if (root.left == null && root.right == null)
                {
                    sum += currSum;
                }
                root = root.right;
            }
            return sum;
        }
        public int SumNumbersWithInBuiltPair(TreeNode root)
        {
            if (root == null)
            {
                return 0;
            }
            int sum = 0, currSum = 0;
            Stack<Tuple<TreeNode, int>> s = new Stack<Tuple<TreeNode, int>>();
            while (root != null || s.Count != 0)
            {
                while (root != null)
                {
                    currSum = currSum * 10 + root.val;
                    s.Push(new Tuple<TreeNode, int>(root, currSum));
                    root = root.left;
                }
                Tuple<TreeNode, int> p = s.Pop();
                root = (TreeNode)p.Item1;
                currSum = (int)p.Item2;
                if (root.left == null && root.right == null)
                {
                    sum += currSum;
                }
                root = root.right;
            }
            return sum;
        }
    }
}

File: Sum Root to Leaf Numbers.cpp
// TC = O(N)
// SC = O(H)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// class Solution {
//     // TC = O(N)
//     // TC = O(H)
//     int result;
//     public int sumNumbers(TreeNode root) {
//         helper(root, 0);
//         return result;
//     }
//     private void helper(TreeNode root, int currSum) {
//         // base
//         if(root == null) return;
//         //logic - at every step we are multiplying by 10 (parent's val) and adding current val and storing it as local sum -> currSum 
//         helper(root.left, currSum * 10 + root.val);
//         helper(root.right, currSum * 10 + root.val);
//         // then we are checking if it a leaf then add the value to result
//         // this can be done anywhere in any style like in/ pre/ postorder
//         if(root.left == null && root.right == null) result += currSum * 10 + root.val;
//     }
// }

// Iterative
class Solution {
    // TC = O(N)
    // TC = O(H)
    public int sumNumbers(TreeNode root) {
        int result = 0, currNum = 0;
        Stack<TreeNode> st = new Stack<>(); // 2 stacks as we are pushing root & currSum in rec st
        Stack<Integer> currNumSt = new Stack<>();
        while(root != null || !st.isEmpty()) {
            while(root != null) {
                currNum = currNum * 10 + root.val;
                st.push(root);
                currNumSt.push(currNum);
                root = root.left;
            }
            //st.pop()
            root = st.pop();
            currNum = currNumSt.pop();
            if(root.left == null && root.right == null) result += currNum; // elaf node
            root = root.right;
        }
        return result;
    }
}